content/en/docs/tasks/access-kubernetes-api/custom-resource-definition-versioning.md

```diff
@@ -0,0 +1,154 @@
+    shortNames:
+    - ct
+```
+
+If you save the above definition to `my-versioned-crontab.yaml` and create it:
+
+```shell
+kubectl create -f my-versioned-crontab.yaml
+```
+
+By creating this CustomResourceDefinition, API server will add endpoints for each enabled version (e.g. /apis/stable.example.com/v1 and /apis/stable.example.com/v2).
+
+## Version order
+The version name in the CustomResourceDefiniton version list is used to compute the order of versions. If the version string is "kube-like", it will sort above non "kube-like" version strings, which are ordered
+lexicographically. "Kube-like" versions start with a "v", then are followed by a number (the major version),
+then optionally the string "alpha" or "beta" and another number (the minor version). These are sorted first
+by GA > beta > alpha (where GA is a version with no suffix such as beta or alpha), and then by comparing
+major version, then minor version. This is an example of sorted versions list:
+
+```
+- v10
+- v2
+- v1
+- v11beta2
+- v10beta3
+- v3beta1
+- v12alpha1
+- v11alpha2
+- foo1
+- foo10.
+```
+
+The order of the version defines a priority for each version. Most importantly the version that comes first in term of priority
+will be used by kubectl as the default version to access objects. For `my-versioned-crontab.yaml` example, the version order would be v2, v1. That means any kubectl command will use v2 as default version unless the version can be infered from the provided object.
+
+## Conversion
+Currently API Server does not do any conversions on CustomResourceDefinition other than fixing apiVersion of the object. The stored object will always have the storage version's apiVersion and return objects will have the same apiVersion as the request version.
+
+If you save the following YAML to `my-crontab-v1.yaml`:
+
+```yaml
+apiVersion: "stable.example.com/v1"
+kind: CronTab
+metadata:
+  name: my-new-cron-object
+spec:
+  cronSpec: "* * * * */5"
+  image: my-awesome-cron-image
+```
+
+and create it:
+
+```shell
+kubectl create -f my-crontab.yaml
+```
+
+it will store the object with `stable.example.com/v1` API version. However when you access the object:
+
+```shell
+kubectl get my-new-cron-object -o yaml
+```
+
+you should see the object with `stable.example.com/v2` API version as `v2` is the prefered version.
+
+```yaml
+apiVersion: stable.example.com/v2
+kind: CronTab
+metadata:
+  clusterName: ""
+  creationTimestamp: 2017-05-31T12:56:35Z
+  deletionGracePeriodSeconds: null
+  deletionTimestamp: null
+  name: my-new-cron-object
+  namespace: default
+  resourceVersion: "285"
+  selfLink: /apis/stable.example.com/v1/namespaces/default/crontabs/my-new-cron-object
+  uid: 9423255b-4600-11e7-af6a-28d2447dc82b
+spec:
+  cronSpec: '* * * * */5'
+  image: my-awesome-cron-image
+```
+
+## Stored Versions
+It is possible to have different versions of Custom Resources in the storage. When changing the storage version, it is
```

Comment:
... in the storage at the same time.

Reply:
this must be much much clearer as this is crucial for not loosing your data:

It is possible to have different versions of Custom Resources in the storage at the same time. When changing the storage version in the CustomResourceDefinition you only affect new and updated objects. There is no automatic upgrade of your persisted data! Existing old object are still stored in their old version. 

The API server records every version which was marked as storage version ever in the status field `storedVersions`. In other words: expect old versioned objects in etcd for all versions in `status.storedVersions`. Note that we may not actually have objects with all versions listed in `storedVersions` stored in etcd, because the list has any version that has the storage flag set at some time, independently whether object were actually written during that time.

When deprecating versions and dropping support, make sure to have a storage upgrade procedure in place: